pkg/labels/oplabels.go

```diff
@@ -0,0 +1,246 @@
+
+	//orchestrationInfo - labels from orchestration which are not used in determining a security identity
+	OrchestrationInfo Labels
+}
+
+// NewOpLabels creates new initialized OpLabels
+func NewOpLabels() OpLabels {
+	return OpLabels{
+		Custom:                Labels{},
+		Disabled:              Labels{},
+		OrchestrationIdentity: Labels{},
+		OrchestrationInfo:     Labels{},
+	}
+}
+
+// SplitUserLabelChanges returns labels to 'add' and 'del'ete to make
+// the custom labels match 'lbls'
+// FIXME: Somewhere in the code we crash if the returned maps are non-nil
+// but length 0. We retain this behaviour here because it's easier.
+func (o *OpLabels) SplitUserLabelChanges(lbls Labels) (add, del Labels) {
+	for key, lbl := range lbls {
+		if _, found := o.Custom[key]; !found {
+			if add == nil {
+				add = Labels{}
+			}
+			add[key] = lbl
+		}
+	}
+
+	for key, lbl := range o.Custom {
+		if _, found := lbls[key]; !found {
+			if del == nil {
+				del = Labels{}
+			}
+			del[key] = lbl
+		}
+	}
+
+	return add, del
+}
+
+// IdentityLabels returns map of labels that are used when determining a
+// security identity.
+func (o *OpLabels) IdentityLabels() Labels {
+	enabled := make(Labels, len(o.Custom)+len(o.OrchestrationIdentity))
+
+	for k, v := range o.Custom {
+		enabled[k] = v
+	}
+
+	for k, v := range o.OrchestrationIdentity {
+		enabled[k] = v
+	}
+
+	return enabled
+}
+
+// GetIdentityLabel returns the value of the given Key from all IdentityLabels.
+func (o *OpLabels) GetIdentityLabel(key string) (l Label, found bool) {
+	l, found = o.OrchestrationIdentity[key]
+	if !found {
+		l, found = o.Custom[key]
+	}
+	return l, found
+}
+
+// AllLabels returns all Labels within the provided OpLabels.
+func (o *OpLabels) AllLabels() Labels {
+	all := make(Labels, len(o.Custom)+len(o.OrchestrationInfo)+len(o.OrchestrationIdentity)+len(o.Disabled))
+
+	for k, v := range o.Custom {
+		all[k] = v
+	}
+
+	for k, v := range o.Disabled {
+		all[k] = v
+	}
+
+	for k, v := range o.OrchestrationIdentity {
+		all[k] = v
+	}
+
+	for k, v := range o.OrchestrationInfo {
+		all[k] = v
+	}
+	return all
+}
+
+func (o *OpLabels) ReplaceInformationLabels(l Labels, logger *logrus.Entry) bool {
+	changed := false
+	keepers := make(keepMarks)
+	for _, v := range l {
+		keepers.set(v.Key)
+		if o.OrchestrationInfo.upsertLabel(v) {
+			changed = true
+			if logger != nil {
+				logger.WithField(logfields.Object, logfields.Repr(v)).Debug("Assigning information label")
+			}
+		}
+	}
+	o.OrchestrationInfo.deleteUnMarked(keepers)
+
+	return changed
+}
+
+func (o *OpLabels) ReplaceIdentityLabels(l Labels, logger *logrus.Entry) bool {
+	changed := false
+
+	keepers := make(keepMarks)
+	disabledKeepers := make(keepMarks)
+
+	for k, v := range l {
+		// A disabled identity label stays disabled without value updates
+		if _, found := o.Disabled[k]; found {
+			disabledKeepers.set(k)
+		} else if keepers.set(v.Key); o.OrchestrationIdentity.upsertLabel(v) {
+			if logger != nil {
```

Comment:
`logger` here is the endpoint's logger passed in as an argument. We do call `UpdateLogger()` on all code paths that create endpoints. though, so it will not be nil. Changed accordingly.